recidiviz/common/constants/enum_overrides.py

```
@@ -0,0 +1,93 @@
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+# ============================================================================
+"""Scraper-specific mappings from strings to enum values."""
+from collections import defaultdict
+from typing import Callable, Dict, Set, Tuple, Optional, Union
+
+from recidiviz.common.constants.entity_enum import EntityEnumMeta, EntityEnum
+
+
+class EnumOverrides:
+    """Contains scraper-specific mappings from string keys to EntityEnum
+    values. EnumOverrides objects should be created using EnumOverrides.Builder.
+    """
+
+    def __init__(self, maps, ignores, lambda_maps):
+        self._maps: Dict[EntityEnumMeta,
+                         Dict[str, EntityEnum]] = maps
+        self._ignores: Dict[Optional[EntityEnumMeta], Set[str]] = ignores
+        self._lambda_maps: Dict[EntityEnumMeta,
+                                Set[Tuple[Callable[[str], bool],
+                                          EntityEnum]]] = lambda_maps
+
+    def should_ignore(self,
+                      k: str,
+                      enum_class: EntityEnumMeta = None) -> bool:
+        return k in self._ignores[enum_class] or k in self._ignores[None]
+
+    def parse(self,
+              k: str,
+              enum_class: EntityEnumMeta) -> Optional[EntityEnum]:
+        if self.should_ignore(k) or self.should_ignore(k, enum_class):
+            return None
+
+        direct_lookup = self._maps[enum_class].get(k)
+        if direct_lookup:
+            return direct_lookup
+
+        for matcher, v in self._lambda_maps[enum_class]:
+            if matcher(k):
+                return v
```

If we want to short circuit here, then we need to ensure that `self._lambda_maps` has a deterministic order (ie. Use `ordereddict` instead of a `dict`).

Personally, I think we should just throw an exception if a provided string matches multiple matchers since it seems very unlikely that a scraper writer would knowingly depend on the order of matchers. Ie, something like this might be possible:
```
matches = [v for matcher, v in self._lambda_maps if matcher(k)]
return more_itertools.one(matches)
```

Thoughts on this?